server/src/test/java/org/elasticsearch/snapshots/mockstore/MockEventuallyConsistentRepository.java

```diff
@@ -0,0 +1,199 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.snapshots.mockstore;
+
+import org.elasticsearch.cluster.coordination.DeterministicTaskQueue;
+import org.elasticsearch.cluster.metadata.RepositoryMetaData;
+import org.elasticsearch.common.blobstore.BlobContainer;
+import org.elasticsearch.common.blobstore.BlobPath;
+import org.elasticsearch.common.blobstore.BlobStore;
+import org.elasticsearch.common.io.Streams;
+import org.elasticsearch.common.xcontent.NamedXContentRegistry;
+import org.elasticsearch.env.Environment;
+import org.elasticsearch.repositories.fs.FsRepository;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.DirectoryNotEmptyException;
+import java.nio.file.FileAlreadyExistsException;
+import java.nio.file.NoSuchFileException;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Mock Repository that simulates the eventually consistent behaviour of AWS S3 as documented in the
+ * <a href="https://docs.aws.amazon.com/AmazonS3/latest/dev/Introduction.html#ConsistencyModel">AWS S3 docs</a>.
+ * Specifically this implementation simulates:
+ * <ul>
+ *     <li>First read after write is consistent for each blob. (see S3 docs for specifics)</li>
+ *     <li>Deletes and updates to a blob can become visible with a delay.</li>
```

Comment:
Can you elaborate on how we're simulating it here?
From what I see, `pendingWrites` map will always contain the latest write in it and `readBlob` executes `ensureReadAfterWrite`, that will run the latest `pendingWrite` if there is no cached miss.
So delays here are emulated for cached misses only, I don't see how the read could read, for example, the first write if the second write is delayed.
Can you please give more precise description of what this implementation is doing?

Reply:
> So delays here are emulated for cached misses only, I don't see how the read could read, for example, the first write if the second write is delayed.

This is not implemented and I chose to add this todo instead: https://github.com/elastic/elasticsearch/pull/40893/files#diff-5c02ffbdee1998c540bcfebd54c4b422R164

There is no point in implementing out of order updates to blobs explained above:
* Updates to blobs have no guaranteed read visibility -> our logic shouldn't rely on updates to blobs ever and we need to remove any spots where they occur and then add an assertion here that simply throws on updates.

So what we have now is simply what is listed there, the fact that two consecutive writes to a blob without a read will result in the second write result showing up consistently is addressed by the todo as well.

Reply:
@original-brownbear Do these statements describe what your implementation is doing more precisely?
1) If there is no cached miss, read always returns the latest write.
2) If there is a cached miss, read might return nothing or the latest write.
3) Writes and deletes become visible to list operations with a delay, but never out of order.

Reply:
@andrershov sounds about the same to me :) Though I'm not so sure about the phrase "cached miss". It's a little iffy calling it that in docs when it's not a phrase used by the AWS S3 docs linked above the text.
Maybe it would make it easier if I just added a line to what is there already that simply states:

"Contrary to what is stated in the AWS S3 documentation, in this implementation writes and deletes become visible to readers in the order they were executed in. A list or get operation will never see an older state after already having seen a newer state."

?